### PR TITLE
Implement hiprtc API for basic runtime compilation cases

### PR DESCRIPTION
**Pull request by linehill**
Implement hiprtc API for basic runtime compilation cases

Supported:

* Invoking extern "C" `__global__` kernels.

Unsupported or missing features:

* Compile options (currently ignored).
* Invoking non-extern "C" `__global__` kernels (required `hiprtcAddNameExpression()` and `hiprtcGetLoweredName()` functions are not implemented yet).

### CODE REVIEW

**Review by pjaaskel — APPROVED** [review]
Great work, Henry. Does this enable some tests to the test suite?

**Comment by linehill** [comment]
> Does this enable some tests to the test suite?

A test is enabled by this commit: 16b2712.

**Comment by pjaaskel** [comment]
@pvelesko is this OK to push? Does it suffice for libceed?

**Comment by pvelesko** [comment]
I will review today or tomorrow - working on HIP submodule refactoring that's required for CHIP-SPV integration into HIPCOMMON.

**Comment by pvelesko** [comment]
```
/home/pvelesko/space/CHIP-SPV/src/Utils.cc:27:10: fatal error: 'filesystem' file not found
```

I need to replace this with `experimental::filesystem` to get past compilation. @linehill

**Comment by linehill** [comment]
<filesystem> should be provided by C++17 standard. AFAIK, we don't require the standard but the root CMakeLists.txt defaults to it:
```
set(CMAKE_CXX_STANDARD 17 CACHE STRING "The C++ standard to use.")
set(CMAKE_CXX_STANDARD_REQUIRED ON)
```

**Comment by pvelesko** [comment]
Can you push the fix? @linehill

I'm in the process of replacing fs with experimental fs but still running into linking issues

**Comment by pvelesko** [comment]
Not sure what you mean by that? I'm using the unmodified CMakeLists.txt

**Comment by linehill** [comment]
<filesystem> is part of C++17. Assuming CMake compiles the CHIP-SPV with c++17 the header should be found. Could you investigate why it doesn't find it in your system?

**Comment by pvelesko** [comment]
> Could you investigate why it doesn't find it in your system?

Ok I'll look into that.

**Comment by linehill** [comment]
Could be caused by old libstdc++ which lacks the header. Clang does not provide it and uses what is found in the system instead, I think.

**Comment by pvelesko** [comment]
Can we merge the fix then? HIPCC uses std::experimental::fs as well so this wouldn't introduce anything new.

**Comment by linehill** [comment]
The fix may fail on newer standard library not having the <experimental/filesystem> header anymore. HIPCC also uses <filesystem> if the system provides it. I'm working on an alternate fix candidate.

File: src/CHIPBackend.cc [hunk 1]
```diff
@@ -1445,6 +1450,17 @@ CHIPQueue *CHIPBackend::findQueue(CHIPQueue *ChipQueue) {
...
 }
 
+bool CHIPBackend::eraseProgram(const CHIPProgram *ToErase) {
+  auto ErasePos =
+      std::remove_if(ChipPrograms_.begin(), ChipPrograms_.end(),
+                     [&](const std::unique_ptr<CHIPProgram> &Prog) -> bool {
+                       return Prog.get() == ToErase;
+                     });
+  bool WasErased = ErasePos != ChipPrograms_.end();
+  ChipPrograms_.erase(ErasePos, ChipPrograms_.end());
```
Comment (pvelesko) [inline src/CHIPBackend.cc:1460]: thread safe?

File: src/CHIPBackend.hh [hunk 2]
```diff
@@ -1850,6 +1886,12 @@ public:
   /* event interop */
   virtual hipEvent_t getHipEvent(void *NativeEvent) = 0;
   virtual void *getNativeEvent(hipEvent_t HipEvent) = 0;
+
+  void addProgram(std::unique_ptr<CHIPProgram> Program) {
+    ChipPrograms_.push_back(std::move(Program));
```
Comment (pvelesko) [inline src/CHIPBackend.hh:1891]: thread safe?

File: include/hip/spirv_hip_fp16.h [hunk 3]
```diff
@@ -39,8 +39,8 @@ THE SOFTWARE.
 typedef _Float16 _hip_f16;
 typedef _Float16 _hip_f16_2 __attribute__((ext_vector_type(2)));
 #else
-typedef api_half _hip_f16;
-typedef api_half2 _hip_f16_2;
+typedef short _hip_f16;
+typedef short _hip_f16_2 __attribute__((ext_vector_type(2)));
```
Comment (pvelesko) [inline include/hip/spirv_hip_fp16.h:43]: I was confused about this part.. Where does api_half come from?

File: src/CHIPBackend.cc [hunk 4]
```diff
@@ -737,6 +734,14 @@ void CHIPDevice::addModule(const std::string *ModuleStr, CHIPModule *Module) {
   this->ChipModules[ModuleStr] = Module;
 }
 
+void CHIPDevice::eraseModule(CHIPModule *Module) {
+  auto *ModSrc = Module->getModuleSource();
+  if (auto It = ChipModules.find(ModSrc); It != ChipModules.end()) {
+    delete It->second;
+    ChipModules.erase(It);
```
Comment (pvelesko) [inline src/CHIPBackend.cc:741]: thread safe?

File: include/hip/spirv_hip_fp16.h (hunk 3 above)
Comment (linehill) [reply include/hip/spirv_hip_fp16.h:43]: No idea where it was supposed to come from but it not defined anywhere, AFAIK.

File: src/CHIPBackend.cc (hunk 1 above)
Comment (linehill) [reply src/CHIPBackend.cc:1460]: I have a follow up PR for removing this. I'll transfer it here.

File: src/CHIPBackend.hh (hunk 2 above)
Comment (linehill) [reply src/CHIPBackend.hh:1891]: I'll remove this.

File: src/Utils.hh [hunk 5]
```diff
@@ -24,9 +24,56 @@
 #ifndef SRC_UTILS_HH
 #define SRC_UTILS_HH
 
+#include "common.hh"
+
+#include "hip/hip_fatbin.h"
+
+#include <filesystem>
+#include <optional>
```
Comment (pvelesko) [inline src/Utils.hh:32]: unnecessary if included in common.hh

**Comment by pvelesko** [comment]
Nice work! I think we should merge #182 first as that will enable a lot more hipRTC tests.

**Comment by linehill** [comment]
I'll rebase this patch when it's merged.

File: src/CHIPBackend.cc (hunk 4 above)
Comment (linehill) [reply src/CHIPBackend.cc:742]: Added device scope lock.

File: src/Utils.hh (hunk 5 above)
Comment (linehill) [reply src/Utils.hh]: Filesystem include is gone. Didn't see optional included in common.h.

**Comment by linehill** [comment]
Updated. The patch tries to use `<experimental/filesystem>` header instead if the `<filesystem>` header is not available.

**Comment by pvelesko** [comment]
Merged the HIPCOMMON rebase. Please checkout HIP/test/README there's a new section about hipRTC testing

**Comment by pvelesko** [comment]
Still waiting on feedback from AMD and the rebased branch needs some additional work so go ahead and merge this when you are ready.

**Comment by linehill** [comment]
Alright, thanks.

**Pull request merged**